includes/class-amp-theme-support.php

```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
 				self::redirect_non_amp_url( current_user_can( 'manage_options' ) ? 302 : 301 );
 			}
 
+			if ( MobileRedirectManager::is_enabled() && is_amp_available() ) {
+				$mobile_redirection_disabled = isset( $_GET[ MobileRedirectManager::NO_AMP_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+				// If `noamp` query var is present, set a cookie to persist disabling of mobile redirection (if one has not already been set).
+				if ( $mobile_redirection_disabled && ! MobileRedirectManager::redirection_disabled() ) {
+					setcookie(
+						MobileRedirectManager::DISABLED_COOKIE_NAME,
+						'1',
+						[
+							'expires'  => 0, // Cookie will only last for the current browser session.
+							'path'     => '/',
+							'secure'   => isset( $_SERVER['HTTPS'] ),
+							'httponly' => false,
+							'samesite' => 'strict',
+						]
+					);
+				}
+
+				// Redirect if `noamp` query parameter is not present and JS redirection is disabled.
+				if ( ! $mobile_redirection_disabled && ! MobileRedirectManager::should_redirect_via_js() ) {
+					$amp_url = add_query_arg( amp_get_slug(), '1', amp_get_current_url() );
+					wp_safe_redirect( $amp_url, current_user_can( 'manage_options' ) ? 302 : 301 );
```

Comment:
I think 302 is sufficient here. By virtue of the cookie being session-based, this entails a temporary redirect. The only reason for doing 301 redirect was for the sake of anyone concerned about SEO. The 302 redirect was used for people who were admins since they can change the mode. But just only doing 302 should be all that is needed here:

```suggestion
					wp_safe_redirect( $amp_url, 302 );
```

Speaking of which, I bet the link to leave the mobile site should have a `rel=nofollow`.